File: day04.hs
#!/usr/bin/env cabal
{- cabal:
build-depends: base
             , trifecta
-}

import System.IO (readFile)
import Text.Trifecta
import Data.Foldable (foldr)
import Control.Monad (mfilter)

type Range = (Integer, Integer)
type Pair = (Range, Range)

range :: Parser Range
range = do
  a <- integer
  char '-'
  b <- integer
  return (a, b)

pair :: Parser Pair
pair = do
  a <- range 
  char ','
  b <- range
  return (a, b)

pairs :: Parser [Pair]
pairs = some pair

contains :: Range -> Range -> Bool
contains (a0, a1) (b0, b1) = a0 <= b0 && b1 <= a1

overlaps :: Range -> Range -> Bool
overlaps (a0, a1) (b0, b1) = (a0 <= b0 && b0 <= a1) || (b0 <= a0 && a0 <= b1)

problem1 :: [Pair] -> Integer
problem1 pairs =
  foldr (\_ acc -> acc + 1) 0 (mfilter (\(a, b) -> contains a b || contains b a) pairs)

problem2 :: [Pair] -> Integer
problem2 pairs =
  foldr (\_ acc -> acc + 1) 0 (mfilter (\(a, b) -> overlaps a b) pairs)

main :: IO ()
main = do
  input <- readFile "day04.txt"
  let thePairs = parseString pairs mempty input
  
  print $ problem1 <$> thePairs
  print $ problem2 <$> thePairs

File: day04.py
from typing import Tuple, Generator, Iterable

Range = Tuple[int, int]

def contains(a: Range, b: Range) -> bool:
    return a[0] <= b[0] and b[1] <= a[1]

def overlaps(a: Range, b: Range) -> bool:
    return (a[0] <= b[0] and b[0] <= a[1]) or \
        (b[0] <= a[0] and a[0] <= b[1])

def parse_range(s: str) -> Range:
    a, b = s.split('-')
    return int(a), int(b)

def read_input(path) -> Generator[Tuple[Range, Range], None, None]:
    with open(path) as f:
        for line in f.readlines():
            fst, snd = line.rstrip().split(',')
            yield parse_range(fst), parse_range(snd)

def problem1(pairs: Iterable[Tuple[Range, Range]]) -> int:
    return sum(
        1
        for a, b in pairs
        if contains(a, b) or contains(b, a)
    )

def problem2(pairs: Iterable[Tuple[Range, Range]]) -> int:
    return sum(
        1 for a, b in pairs if overlaps(a, b)
    )

if __name__ == '__main__':

    pairs = list(read_input('day04.txt'))

    print(problem1(pairs))
    print(problem2(pairs))



